aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -1565,41 +1517,53 @@ Tensor matmul(
                                          .reshape({expand_batch_product, n, m1});
     const auto tensor2_expanded = tensor2.expand(tensor2_expand_size)
                                          .reshape({expand_batch_product, m2, p});
+    output_shape.append({n, p});
 
-    // reshape batches back into result
-    // we make the most likely path explicit
-    if C10_LIKELY (dim_tensor1 > 1) {
-      if C10_LIKELY (dim_tensor2 > 1) {
-        output_shape.append({n, p});
-      } else {
-        output_shape.push_back(n);
-      }
-    } else { // dim_tensor2 > 1
-      output_shape.push_back(p);
-    }
+    result.set_(at::_unsafe_view(at::bmm_out(result, tensor1_expanded, tensor2_expanded),output_shape));
```

Comment:
Is `set_` really a valid thing to use in out variants? Here the result might be already of the right size and so as per the out= contract, you should write inplace into its memory.

Reply:
It is writing in place into its memory in `bmm_out`, but then we do all the `_unsafe_view` business to reshape it in case we flattened from, say, shape `(4, 3, 2, 1)` into shape `(12, 2, 1)`.

This code was not added in this PR and it's currently used in `matmul`, so it better be correct...